jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+            Entry entry = queue.poll();
+            if (entry != null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} resuming {}", this, entry.request);
+                // Try again to acquire a semaphore pass.
+                // If it cannot be acquired, the request may
+                // be re-prioritized with a higher priority.
+                processAndComplete(entry.request, entry.response, entry.callback);
+                return;
+            }
+        }
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} no suspended requests to resume", this);
+    }
+
+    private class Entry implements CyclicTimeouts.Expirable
+    {
+        private final Request request;
+        private final Response response;
+        private final Callback callback;
+        private final int priority;
+        private final long expireNanoTime;
+
+        private Entry(Request request, Response response, Callback callback, int priority)
+        {
+            this.request = request;
+            this.response = response;
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : NanoTime.now() + maxSuspend.toNanos();
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return expireNanoTime;
+        }
+
+        private void expire()
+        {
+            // The request timed out, therefore it never acquired a semaphore pass.
+            boolean removed = queues.get(priority).remove(this);
+            if (removed)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} timeout {}", QoSHandler.this, request);
+                Response.writeError(request, response, callback, HttpStatus.SERVICE_UNAVAILABLE_503);
+            }
+        }
+    }
+
+    private class Resumer extends HttpStream.Wrapper
+    {
+        private final Request request;
+
+        private Resumer(HttpStream wrapped, Request request)
+        {
+            super(wrapped);
+            this.request = request;
+        }
+
+        @Override
+        public void succeeded()
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} succeeded {}", QoSHandler.this, request);
+            resume();
+        }
+
+        @Override
+        public void failed(Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} failed {}", QoSHandler.this, request, x);
+            resume();
+        }
+    }
+
+    private class Timeouts extends CyclicTimeouts<Entry>
+    {
+        private Timeouts(Scheduler scheduler)
+        {
+            super(scheduler);
+        }
+
+        @Override
+        protected Iterator<Entry> iterator()
+        {
+            return queues.stream()
+                .flatMap(Queue::stream)
+                .iterator();
```

Comment:
Can we avoid streams?  I know it will be a bit more code, but this iterator is called frequently

Reply:
Ok, but can you say that in a comment so we don't forget